File: RIS_near_field_beamforming/near_field_RIS.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <complex>
#include <iomanip>
#include <cstdint>
#include <omp.h>

#define PI 3.14159265358979323846

struct Params {
    double lambda;
    double dx, dy;
    int M, N;
    double Pt;
    double Gt_dB, Gr_dB;
    double G;
    double A;
    double d1;
    double x_rx, y_rx, z_rx;
};

double db2lin(double db) {
    return std::pow(10.0, db / 10.0);
}

double distance3D(double x1, double y1, double z1,
                  double x2, double y2, double z2) {
    return std::sqrt((x1 - x2)*(x1 - x2) +
                     (y1 - y2)*(y1 - y2) +
                     (z1 - z2)*(z1 - z2));
}

// Compute Pr for a given binary pattern
double compute_Prx(const std::vector<std::vector<int>>& binary, const Params& p) {
    std::complex<double> sum = 0.0;
    double k = 2.0 * PI / p.lambda;
    double z_tx = -p.d1;

    for (int n = 0; n < p.N; ++n) {
        for (int m = 0; m < p.M; ++m) {
            double x_nm = (m - (p.M - 1) / 2.0) * p.dx;
            double y_nm = (n - (p.N - 1) / 2.0) * p.dy;
            double z_nm = 0.0;

            double rt = distance3D(x_nm, y_nm, z_nm, 0, 0, z_tx);
            double rr = distance3D(x_nm, y_nm, z_nm, p.x_rx, p.y_rx, p.z_rx);
            double total_path = rt + rr;

            double phase_shift = binary[n][m] == 1 ? PI : 0.0;
            std::complex<double> exp_term = std::exp(std::complex<double>(0, -1) * (k * total_path - phase_shift));
            sum += exp_term / (rt * rr);
        }
    }

    double Gt = db2lin(p.Gt_dB);
    double Gr = db2lin(p.Gr_dB);
    double prefactor = (p.Pt * Gt * Gr * p.G * p.dx * p.dy * p.lambda * p.lambda * p.A * p.A) / (64.0 * std::pow(PI, 3));
    return prefactor * std::norm(sum);
}

int main() {
    Params p;

    // System configuration
    p.lambda = 0.085;
    p.dx = p.dy = 0.015;
    p.M = p.N = 16;
    p.Pt = 1e-3;
    p.Gt_dB = 60;
    p.Gr_dB = 34;
    p.G = 8.0;
    p.A = 1.0;
    p.d1 = 3.5;

    // Target Rx location
    double d2 = 1.0;
    double azimuth_deg = 30.0;
    double azimuth_rad = azimuth_deg * PI / 180.0;
    p.x_rx = d2 * std::cos(azimuth_rad);
    p.y_rx = d2 * std::sin(azimuth_rad);
    p.z_rx = 0.0;

    uint64_t total_patterns = 65536;
    double max_power = -1.0;
    double min_power = 1e9;
    uint16_t best_pattern = 0;
    uint16_t worst_pattern = 0;

    #pragma omp parallel for
    for (uint64_t i = 0; i < total_patterns; ++i) {
        std::vector<std::vector<int>> binary(p.N, std::vector<int>(p.M, 0));

        // Build binary RIS matrix from column bitmask
        for (int m = 0; m < p.M; ++m) {
            int on = (i >> m) & 1;
            for (int n = 0; n < p.N; ++n) {
                binary[n][m] = on;
            }
        }

        double Pr = compute_Prx(binary, p);

        #pragma omp critical
        {
            if (Pr > max_power) {
                max_power = Pr;
                best_pattern = i;
            }
            if (Pr < min_power) {
                min_power = Pr;
                worst_pattern = i;
            }
        }
    }

    // Output best pattern
    std::cout << "\n=== BEST PATTERN ===\n";
    std::cout << "Bitmask: 0x" << std::hex << best_pattern << std::dec << "\nColumns: ";
    for (int m = 0; m < p.M; ++m) {
        std::cout << ((best_pattern >> m) & 1) << " ";
    }
    std::cout << "\nMax Power: " << std::scientific << max_power << " W\n";

    // Output worst pattern
    std::cout << "\n=== WORST PATTERN ===\n";
    std::cout << "Bitmask: 0x" << std::hex << worst_pattern << std::dec << "\nColumns: ";
    for (int m = 0; m < p.M; ++m) {
        std::cout << ((worst_pattern >> m) & 1) << " ";
    }
    std::cout << "\nMin Power: " << std::scientific << min_power << " W\n";

    return 0;
}

File: RIS_near_field_beamforming/near_field_RIS.py
import numpy as np
import math
from concurrent.futures import ThreadPoolExecutor

# Constants
PI = np.pi

# Parameters
class Params:
    def __init__(self):
        self.lambda_ = 0.085
        self.dx = self.dy = 0.026
        self.M = self.N = 16
        self.Pt = 1e-3
        self.Gt_dB = 20
        self.Gr_dB = 20
        self.G_dB = 15
        self.A = 0.8
        self.d1 = 3.0

        azimuth_deg = -40.0
        azimuth_rad = np.radians(azimuth_deg)
        d2 = 3.0
        self.x_rx = d2 * np.cos(azimuth_rad)
        self.y_rx = d2 * np.sin(azimuth_rad)
        self.z_rx = 0.0

def db2lin(db):
    return 10 ** (db / 10)

def distance3D(x1, y1, z1, x2, y2, z2):
    return np.sqrt((x1 - x2)**2 + (y1 - y2)**2 + (z1 - z2)**2)

def compute_Prx(binary, p):
    k = 2 * PI / p.lambda_
    z_tx = p.d1
    total_field = 0 + 0j

    for m in range(p.N):
        for n in range(p.M):
            x_m = (m - (p.M - 1) / 2.0) * p.dx
            y_n = (n - (p.N - 1) / 2.0) * p.dy

            rt = distance3D(x_m, y_n, 0, 0, 0, z_tx)
            rr = distance3D(x_m, y_n, 0, p.x_rx, p.y_rx, p.z_rx)
            total_path = rt + rr

            phase = PI if binary[n][m] == 1 else 0
            if phase !=0:
                azimuth_nm=((k*total_path)%(2*phase))
            else:
                azimuth_nm = 0
            exp_term = np.exp(-1j * (k * total_path - azimuth_nm))
            total_field += exp_term / (rt * rr)

    Gt = db2lin(p.Gt_dB)
    Gr = db2lin(p.Gr_dB)
    G = db2lin(p.G_dB)
    prefactor = (p.Pt * Gt * Gr * G * p.dx * p.dy * p.lambda_**2 * p.A**2) / (64 * PI**3)
    return prefactor * abs(total_field)**2

def generate_pattern_from_int(pattern_int, M, N):
    binary = np.zeros((N, M), dtype=int)
    for m in range(M):
        if (pattern_int >> m) & 1:
            binary[:, m] = 1
    return binary

def convert(binary):
    # patterns = []
    # binary = bin(i)[2:].zfill(16)
    hex_pattern = "!0X"
    for bit in binary:
        hex_pattern += "FFFF" if bit == 1 else "0000"
    # patterns.append(hex_pattern)
    return hex_pattern

def generate_256bit_hex_pattern(binary_matrix):
    pattern_256 = 0
    for row in range(16):
        row_bits = 0
        for bit in range(16):
            row_bits |= (binary_matrix[row][bit] << (15 - bit))
        pattern_256 = (pattern_256 << 16) | row_bits
    return f"0x{pattern_256:064X}"

def evaluate_pattern(i, p):
    binary = generate_pattern_from_int(i, p.M, p.N)
    power = compute_Prx(binary, p)
    return i, power

def format_pattern(pattern, M):
    return [(pattern >> m) & 1 for m in range(M)]

# Main execution
if __name__ == "__main__":
    p = Params()
    best_patterns = []
    worst_patterns = []

    total_patterns = 2**16
    print("Evaluating 65,536 column patterns...")

    with ThreadPoolExecutor() as executor:
        results = executor.map(lambda i: evaluate_pattern(i, p), range(total_patterns))

        for pattern, power in results:
            # Insert into top 10 best
            if len(best_patterns) < 10 or power > best_patterns[-1]['power']:
                binary = generate_pattern_from_int(pattern, p.M, p.N)
                hex256 = generate_256bit_hex_pattern(binary)
                entry = {
                    "pattern": pattern,
                    "power": power,
                    "binary": binary,
                    "hex256": hex256
                }
                best_patterns.append(entry)
                best_patterns = sorted(best_patterns, key=lambda x: -x["power"])[:10]

            # Insert into top 10 worst
            if len(worst_patterns) < 10 or power < worst_patterns[-1]['power']:
                binary = generate_pattern_from_int(pattern, p.M, p.N)
                hex256 = generate_256bit_hex_pattern(binary)
                entry = {
                    "pattern": pattern,
                    "power": power,
                    "binary": binary,
                    "hex256": hex256
                }
                worst_patterns.append(entry)
                worst_patterns = sorted(worst_patterns, key=lambda x: x["power"])[:10]

    # Display Best 10
    print("\n=== TOP 10 BEST PATTERNS ===")
    for i, item in enumerate(best_patterns):
        # print(f"\nRank {i+1}")
        # print(f"Bitmask (16-bit): 0x{item['pattern']:04X}")
        print("Columns ON/OFF   :", format_pattern(item["pattern"], p.M))
        print(convert(format_pattern(item["pattern"], p.M)))
        # print("256-bit Hex Pattern:", item["hex256"])
        print(f"Power            : {item['power']:.3e} W ({10 * np.log10(item['power']):.2f} dBW)")

    # Display Worst 10
    # print("\n=== TOP 10 WORST PATTERNS ===")
    # for i, item in enumerate(worst_patterns):
    #     print(f"\nRank {i+1}")
    #     print(f"Bitmask (16-bit): 0x{item['pattern']:04X}")
    #     print("Columns ON/OFF   :", format_pattern(item["pattern"], p.M))
    #     print("256-bit Hex Pattern:", item["hex256"])
    #     print(f"Power            : {item['power']:.3e} W ({10 * np.log10(item['power']):.2f} dBW)")



